### DISCUSSION TITLE
How can one run childrenToReact before component is rendered by React?

### DISCUSSION THREAD

**Discussion by sunknudsen** [remarkjs/remark#1109] [Q&A]
Trying to compute heading slugs programatically… but `element` only runs when component is rendered by React so `toc` is always empty.

```tsx
import githubSlugger from "github-slugger"
import { html } from "property-information"
import { createElement, Fragment, ReactNode } from "react"
import { childrenToReact } from "react-markdown/lib/ast-to-react"
import reactNodeToString from "react-node-to-string"
import rehypeRaw from "rehype-raw"
import remarkParse from "remark-parse"
import remarkRehype from "remark-rehype"
import { unified } from "unified"
import { VFile } from "vfile"

interface Generated {
  element: ReactNode
  toc: string[]
}

export const generate = (content: string): Generated => {
  const processor = unified()
    .use(remarkParse)
    .use(remarkRehype, {
      allowDangerousHtml: true,
    })
    .use([rehypeRaw])

  const file = new VFile()

  file.value = content

  const node = processor.runSync(processor.parse(file), file)

  const toc: string[] = []

  const slugger = new githubSlugger()

  const element = createElement(
    Fragment,
    {},
    childrenToReact(
      {
        listDepth: 0,
        options: {
          components: {
            h1: (props) => {
              toc.push(slugger.slug(reactNodeToString(props.children)))
              return <h1>{props.children}</h1>
            },
          },
        },
        schema: html,
      },
      node
    )
  )

  return {
    element,
    toc,
  }
}
```

**Comment by wooorm** [comment]
Hi!

First, you’re using internals of a package. That’s going to break at any time. Undocumented hidden internals.
So the quick answer is: don’t do this!!!!!!

Also, please spend more time rubber duck debugging and explaining your questions.
I’m fine with helping you, but nobody is going to explain how to use hidden APIs.
Ask question about what you want to do, what input you have, what output you want, what you tried, etc!

---

Second, you’re mixing several things. unified and friends give you an AST. You can use that AST!
But you don’t, you turn it into React.
You can render things with React, and then you also get sort of an AST, which you are trying to access. But you don’t render with React here, so you don’t have anything!

Please, can you explain *what* you want to do, and *why* you have all these tools? Why use `rehype-raw`? Why use `react-node-to-string`? Why use internals?
Then I can help you.

**Reply by sunknudsen** [reply]
Thanks for helping out @wooorm.

I absolutely love [react-markdown](https://github.com/remarkjs/react-markdown) but I wasn’t able to achieve my goal (likely because I don’t fully understand how package works).

I am trying to generate a table of contents (as a separate React component, above example is simplified) from heading tags programatically but given [components](https://github.com/remarkjs/react-markdown#api) only run when `<ReactMarkdown>` is rendered, `toc` is always empty.

I would love to not use hidden internals… would you happen to have wisdom to share?

If above use case isn’t clear enough, I can try to expand.

**Reply by wooorm** [reply]
Do you need to evaluate React components (so custom things that change things), and then serialize those as a string? What sort of components do you have?

**Reply by wooorm** [reply]
And why was `rehype-raw` in there?

**Reply by sunknudsen** [reply]
The most simplified explanation of my goal is I am looking for a way to hook into `<ReactMarkdown>`’s  `h1` component to add in-page anchor references and build a `<TableOfContents>` component deterministically.

Before frankensteining above, I used to parse markdown separately to generate TOC but that can lead to inconsistencies. `react-node-to-string` is used to extract string from `h1` children which is fed to `github-slugger`.

> Do you need to evaluate React components (so custom things that change things), and then serialize those as a string? What sort of components do you have?

I don’t understand question… hopefully above gives context. Btw, blessed you joined conversation… I spent all day yesterday trying to figure this out. 🙌

**Reply by wooorm** [reply]
react-markdown is a wrapper around unified, so that users who don’t know much about unified, can more easily use a couple plugins, to turn their markdown into a react tree. It’s meant for more beginners, who want to solve a couple common cases.

unified is a pipeline for parsing some content, transforming and inspecting it, and then turning it into something else.

In this case, you want to do a pretty simple query of the content. You don’t need to compile it to something else. You’re only inspecting the syntax tree.
Pretty simple, because apparently you do have raw HTML strings in there. It would be better if you don’t use raw strings of HTML in markdown. It’s slow.

So, let’s move a level under unified: to the syntax tree utilities: mdast and hast:

```js
import {fromMarkdown} from 'mdast-util-from-markdown'
import {toHast} from 'mdast-util-to-hast'
import {raw} from 'hast-util-raw'
import {selectAll} from 'hast-util-select'
import {toString} from 'hast-util-to-string'
import GithubSlugger from 'github-slugger'

// Parse a string of markdown into a markdown syntax tree:
const mdast = fromMarkdown('Your string of markdown')

// Turn the markdown syntax tree into an HTML syntax tree, keeping raw strings of HTML:
const hastDirty = toHast(mdast, {allowDangerousHtml: true})

// Parse the HTML syntax tree *and those raw strings of HTML*.
const hast = raw(hastDirty)

// Create a new GH slugger.
const slugger = new GithubSlugger()

// Select all `<h1>`s
const headings = selectAll('h1', hast)

// Serialize the text in each `h1`, pass it through `slugger`, yields the results.
const toc = headings.map((d) => slugger.slug(toString(d)))

console.log(toc)

**Reply by wooorm** [reply]
> that can lead to inconsistencies

Which ones?

> I don’t understand question…

To phrase another way: you could have a component in there that renders the current time stamp. Or other dynamically generated things. If you do such things, it’s going to get very complex. You likely don’t though

**Reply by sunknudsen** [reply]
That is promising!

I was (perhaps naively) hoping to add `github-slugger` in-page anchor references within `h1` component while also generating TOC deterministically using same React heading components and slugs.

Remaining in React-land would allow me to do some React magic using [Mantine](https://mantine.dev/)’s [use-scroll-into-view](https://mantine.dev/hooks/use-scroll-into-view/).

Is that somehow possible?

**Reply by wooorm** [reply]
a) `github-slugger` uses state, React components have to be stateless. React would rerender the tree many times, you can’t use the slugger in your replacement components anyway, feel free to try
b) you don’t need `github-slugger` in components. You don’t need to “stay in react-land” to use the things you ask for, use `rehype-slug` in `react-markdown` and you’ll get those IDs

**Reply by sunknudsen** [reply]
Interesting… how would I extract `rehype-slug` heading level, slugs and strings to generate table of contents?

I was hoping to do both using a singular flow to avoid generating TOC separately from markdown.

That said, perhaps it simply isn’t good practice or possible at all.

**Reply by wooorm** [reply]
You don’t need to. `rehype-slug` will add `id` attributes to all headings.
The above code I posted will find the **exact same headings/slugs**.

> I was hoping to do both using a singular flow to avoid generating TOC separately from markdown.

Well, you started with an entire separate flow of how things should go 😅
So you are *also* using `react-markdown`?
It’s quite correct that that’s duplicate work. It’s almost verbatim duplicate. And that insight gives us the solution:

- mdast-util-from-markdown is equivalent to remark-parse (used inside react-markdown)
- mdast-util-to-hast is equivalent to remark-rehype (used inside react-markdown)
- hast-util-raw is equivalent to rehype-raw (which you use)
- the 3 lines from slugger to toc are what you want to achieve. Which is very close to `rehype-slug`.

So let’s take a look at how that works: https://github.com/rehypejs/rehype-slug/blob/1c61632f1526c6b59c80530ed334d4a094c7a091/index.js.
Some extra code that you don’t need. You could copy/paste that and change it a bit:

```diff
+const toc = []

+export default function myFunkyRehypSlugLikePlugin(options = {}) {
-export default function rehypeSlug(options = {}) {

  ...

      if (headingRank(node) && node.properties && !hasProperty(node, 'id')) {
        node.properties.id = prefix + slugs.slug(toString(node))
+       toc.push(node.properties.id)
      }

  ...
```

Now, of course, React is going to rerender your things. Which will call the plugins again. So `toc` will grow and grow. You’ll need to figure something out to fix that. Maybe something like.

```diff
 export default function myFunkyRehypSlugLikePlugin(options = {}) {
+  toc.length = 0
```

Even with that, I don’t think React will like you for changes state in renders.

**Reply by sunknudsen** [reply]
Wouldn’t using above custom plugin (love the idea and learning a ton btw) always yield an empty `toc` array given `toc` would only be populated once `<ReactMarkdown>` component has rendered?

Looking to render both `<TOC>` and `<ReactMarkdown>` components concurrently.

**Reply by wooorm** [reply]
Well, you are using `react-markdown`, it will render, and then it will populate `toc`.
This was in response to “I was hoping to do both using a singular flow to avoid generating TOC separately from markdown.”
If you don’t want that, there’s my first proposal

**Reply by sunknudsen** [reply]
Thanks for helping out… `myFunkyRehypSlugLikePlugin` idea is what I was looking for. ❤️

**Reply by wooorm** [reply]
Cool :)

**Reply by sunknudsen** [reply]
Would someone happen to know TypeScript type of [tree](https://github.com/rehypejs/rehype-slug/blob/1c61632f1526c6b59c80530ed334d4a094c7a091/index.js#L28)?

**Reply by wooorm** [reply]
That’s inferred. You define it with the `Root` in `Plugin<...>`: https://github.com/rehypejs/rehype-slug/blob/1c61632f1526c6b59c80530ed334d4a094c7a091/index.js#L2.
